pvlib/mlfm.py

```diff
@@ -0,0 +1,804 @@
+        )
+
+        # correction factor
+        corr = (inv_ff - prod) / (inv_ff - tot)
+
+        # put mlfm values in a stack from pr_dc (bottom) to 1/ff_ref (top)
+        # accounting for series and shunt resistance losses
+        dstack['pr_dc'] = +dnorm['pr_dc']  # initialise
+        dstack['i_sc'] = -(dnorm['i_sc'] - 1) * corr
+        dstack['r_sc'] = -(dnorm['r_sc'] - 1) * corr
+        dstack['i_ff'] = -(dnorm['i_ff'] - 1) * corr - gap/2
+        dstack['i_v'] = gap
+        dstack['v_ff'] = -(dnorm['v_ff'] - 1) * corr - gap/2
+        dstack['r_oc'] = -(dnorm['r_oc'] - 1) * corr
+        dstack['v_oc'] = -(dnorm['v_oc'] - 1) * corr
+        dstack['temp_module_corr'] = (
+            -(dnorm['v_oc'] - dnorm['v_oc_temp_corr']) * corr)
+
+        return dstack
+
+    # subtractive losses without series and shunt resistance effects
+    # find factor to transform multiplicative to subtractive losses
+    # correction factor to scale losses to keep 1/ff --> pr_dc
+
+    prod = inv_ff * (
+        dnorm['i_sc'] * dnorm['i_mp'] *
+        dnorm['v_mp'] * dnorm['v_oc']
+    )
+
+    tot = inv_ff + (
+        dnorm['i_sc'] + dnorm['i_mp'] +
+        dnorm['v_mp'] + dnorm['v_oc'] - 4
+    )
+
+    corr = (inv_ff - prod) / (inv_ff - tot)
+
+    # put mlfm values in a stack from pr_dc (bottom) to 1/ff_ref (top)
+    dstack['pr_dc'] = + dnorm['pr_dc']  # initialise
+    dstack['i_sc'] = -(dnorm['i_sc'] - 1) * corr
+    dstack['i_mp'] = -(dnorm['i_mp'] - 1) * corr - gap/2
+    dstack['i_v'] = gap
+    dstack['v_mp'] = -(dnorm['v_mp'] - 1) * corr - gap/2
+    dstack['v_oc'] = -(dnorm['v_oc'] - 1) * corr
+
+    dstack['temp_module_corr'] = (
+        - (dnorm['v_oc'] - dnorm['v_oc_temp_corr']) * corr)
+
+    return dstack
+
+
+def mlfm_6(dmeas, c_1, c_2, c_3, c_4, c_5=0., c_6=0.):
+    r'''
+    Predict normalised LFM values from data in ``dmeas``.
+
+    The normalized LFM values are given by
+
+    .. math::
+
+        c_1 + c_2 (T_m - 25) + c_3 \log10(G_{POA}) + c_4 G_{POA}
+        + c_5 WS + c_6 / G_{POA}
+
+    where :math:`G_{POA}` is global plane-of-array (POA) irradiance in kW/m2,
+    :math:`T_m` is module temperature in C and :math:`WS` is wind speed in
+    m/s.
+
+    Parameters
+    ----------
+    dmeas : DataFrame
+        Must include columns:
+
+        * `'poa_global'` global plane of array irradiance. [W/m^2]
+        * `'temp_module'` module temperature. [C]
+
+        May include optional column:
+
+        * `'wind_speed'` wind speed [m/s].
+
+    c_1 : float
+        Constant term in model.
+    c_2 : float
+        Temperature coefficient in model. [1/C]
+    c_3 : float
+        Coefficient for low light log irradiance drop.
+    c_4 : float
+        Coefficient for high light linear irradiance drop.
+    c_5 : float, default 0
+        Coefficient for wind speed dependence.
+    c_6 : float, default 0
+        Coefficient for dependence on inverse irradiance.
```

Comment:
Looking closer, I think it could be that G_{POA} in the math equation is actually the POA irradiance normalized by `G_STC`, and does NOT actually have units kW/m^2, which means coefficients `c4` and `c6` are indeed unitless. However, I think `c_5` should have units of s/m.

Reply:
Maybe an easier way to look at this is to use the dimensionless value "suns" where [G=1000W/m^2 / G_STC] is 1sun.
This way all the coefficients can be written as "dimensioned"
c_1 is constant,
c_2 is 1/K,
c_3 is /sun,
c_4 is /sun,
c_5 is 1/(ms^-1) as you said
c_6 is *sun,

The values are meaningful in that the
c_3 is the low light "v_oc and r_sh drop" at G=0.1sun (if we use log10 then log10(0.1)=-1)

c_4 is the high light rseries drop at 1sun due to I^2.Rseries

The graph shows the LFM's c_3 and c_4 coefficients quantifiably change the drops at 0.1sun and 1.0sun respectively

![image](https://user-images.githubusercontent.com/13416507/196159656-3495fd1f-ff4a-4d1c-9454-a3582e49fe9e.png)

Reply:
I personally do often think in terms of suns, but I don't think we need to add "suns" units here.

Rather, in the code's documentation I'm seeing some conflation between an irradiance in units of kW/m^2 and a unit-less normalized irradiance that's been divided by G_STC = 1000 W/m^2. Those are significantly different concepts, even if numerically they end up being the same.